### PR TITLE
Version 1.3.2 2023 Tree Cover Loss Update

### PR DESCRIPTION
**Pull request by melrose-wri**
Please check if your PR fulfills the following requirements:
- [x] Make sure you are requesting to pull a topic/feature/bugfix branch (right side). Don't request your master!
- [x] Make sure you are making a pull request against the develop branch (left side). Also you should start your branch off our develop.
- [x] Check the commit's or even all commits' message styles matches our requested structure.
- [x] Check your code additions will fail neither code linting checks nor unit test.

## Pull request type
Please check the type of change your PR introduces:
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 

## What is the current behavior?
Previously, the model used tree cover loss from from 2000 to 2022. 

## What is the new behavior?
The primary update is that the model now uses tree cover loss through 2023. The Global Warming Potential values for CH4 and N2O from peat drainage and burning have also been updated according to IPCC AR6. Also, mp_prep_other_inputs_annual.py and mp_prep_other_inputs_one_off.py have been modified so that specific data preparation steps can be run separately (or all at once) by supplying a command line argument. 

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: data_prep/mp_prep_other_inputs_annual.py [hunk 1]
```diff
@@ -38,140 +43,194 @@ def mp_prep_other_inputs(tile_id_list):
...
     arcpy.management.ProjectRaster("TCL_DD_2022_20230407.tif", r"C:\GIS\raw_data\TCL_DD_2022_20230407_wgs84.tif", 
     'GEOGCS["GCS_WGS_1984",DATUM["D_WGS_1984",SPHEROID["WGS_1984",6378137.0,298.257223563]],PRIMEM["Greenwich",0.0],
     UNIT["Degree",0.0174532925199433]]', "NEAREST", "0.005 0.005", None, None, 'PROJCS["WGS_1984_Goode_Homolosine",
     GEOGCS["GCS_unknown",DATUM["D_WGS_1984",SPHEROID["WGS_1984",6378137.0,298.257223563]],PRIMEM["Greenwich",0.0],
     UNIT["Degree",0.0174532925199433]],PROJECTION["Goode_Homolosine"],PARAMETER["False_Easting",0.0],
     PARAMETER["False_Northing",0.0],PARAMETER["Central_Meridian",0.0],PARAMETER["Option",1.0],UNIT["Meter",1.0]]', "NO_VERTICAL")
-
+    
+    2023 TCL Update: 
+    arcpy.management.ProjectRaster("Goode_FinalClassification_2024_v20240402.tif", r"C:\GIS\carbon_model\Goode_FinalClassification_2023_wgs84_v20240402.tif",
```
Comment (dagibbs22) [inline data_prep/mp_prep_other_inputs_annual.py:58]: Thanks for updating this with the ArcPy code block from the corrected drivers pre-pre-processing.

File: data_prep/mp_prep_other_inputs_one_off.py [hunk 2]
```diff
@@ -5,11 +5,23 @@
...
+4) europe_forest: Creates European natural forest removal rate and standard deviation tiles
+5) us_forest: Creates US forest age, group, and region category tiles
+6) ifl_primary: Creates intact forest landscape/ primary forest tiles
+7) agb_bgb: Creates Hansen tiles of AGB:BGB based on Huang et al. 2021
+8) sdptv2: Creates 6 SDPTv2 tile sets from gfw-data-lake bucket to gfw2-data: AGC removal factors, AGC standard deviations, AGC+BGC removal factors, AGC+BGC standard deviations, planted forest type, and planted forest established year
+
+#TODO: Make the process command line to individually launch all of the pre-processing steps parallelized
+#TODO: Update constants_and_names.py with datalake_pf_estab_year_dir and uncomment 4 lines in sdptv2 section
+#TODO: Plantations gets stuck during the unzip process (line 172)
+#TODO: AGB_BGB netcdf download process fails (line 527)
```
Comment (dagibbs22) [inline data_prep/mp_prep_other_inputs_one_off.py:24]: Thanks for including these TODOs

File: data_prep/mp_prep_other_inputs_annual.py [hunk 3]
```diff
@@ -38,140 +43,194 @@ def mp_prep_other_inputs(tile_id_list):
...
+                 tile_id_list)
+        pool.close()
+        pool.join()
+
+        for output_pattern in [
+            cn.pattern_drivers
+        ]:
+
+            if cn.count == 96:
+                processes = 50  # 60 processors = >730 GB peak (for European natural forest forest removal rates); 50 = XXX GB peak
```
Comment (dagibbs22) [inline data_prep/mp_prep_other_inputs_annual.py:121]: Delete the comment. I doubt that the note on 60 processors for Europe is correct, and it's not relevant for this driver processing call.

File: data_prep/mp_prep_other_inputs_annual.py [hunk 4]
```diff
@@ -38,140 +43,194 @@ def mp_prep_other_inputs(tile_id_list):
...
+        pool.map(partial(uu.mp_warp_to_Hansen, source_raster=source_raster, out_pattern=out_pattern, dt=dt), tile_id_list)
+        pool.close()
+        pool.join()
+
+        for output_pattern in [
+            cn.pattern_TCLF_processed
+        ]:
+
+            if cn.count == 96:
+                processes = 50  # 60 processors = >730 GB peak (for European natural forest forest removal rates); 50 = XXX GB peak
```
Comment (dagibbs22) [inline data_prep/mp_prep_other_inputs_annual.py:208]: Same about deleting the comment about the processors here (and elsewhere it seems copied from some other usage).

File: data_prep/mp_prep_other_inputs_one_off.py [hunk 5]
```diff
@@ -161,6 +239,51 @@ def mp_prep_other_inputs(tile_id_list, process):
...
+    ##################################################################################################
+
+    if process == 'europe_forest' or process == 'all':
+
+        uu.print_log('Pre-processing European natural forest removal rate and standard deviation tiles')
+
+        # List of output directories and output file name patterns
+        output_dir_list = [
+            cn.annual_gain_AGC_BGC_natrl_forest_Europe_dir,
+            #cn.stdev_annual_gain_AGC_BGC_natrl_forest_Europe_dir   # raw files deleted
```
Comment (dagibbs22) [inline data_prep/mp_prep_other_inputs_one_off.py:264]: Any idea why the Europe stdev output is commented out? I probably did that. Not likely to need to rerun, but please comment the stdev output back in unless I'm missing something here.

File: data_prep/mp_prep_other_inputs_one_off.py [hunk 6]
```diff
@@ -169,63 +292,79 @@ def mp_prep_other_inputs(tile_id_list, process):
...
+            output_dir_list = uu.alter_dirs(sensit_type, output_dir_list)
+            output_pattern_list = uu.alter_patterns(sensit_type, output_pattern_list)
+
+        # A date can optionally be provided by the full model script or a run of this script.
+        # This replaces the date in constants_and_names.
+        # Only done if output upload is enabled.
+        if cn.RUN_DATE is not None and cn.NO_UPLOAD is not None:
+            output_dir_list = uu.replace_output_dir_date(output_dir_list, cn.RUN_DATE)
+
+        # Files to process: climate zone, IDN/MYS plantations before 2000, tree cover loss drivers, combine IFL and primary forest
```
Comment (dagibbs22) [inline data_prep/mp_prep_other_inputs_one_off.py:364]: Update this list of files to process.

File: data_prep/mp_prep_other_inputs_one_off.py [hunk 7]
```diff
@@ -617,4 +840,4 @@ def mp_prep_other_inputs(tile_id_list, process):
     # Checks whether the tile_id_list argument is valid
     tile_id_list = uu.tile_id_list_check(tile_id_list)
 
-    mp_prep_other_inputs(tile_id_list=tile_id_list, process=int(process))
+    mp_prep_other_inputs(tile_id_list=tile_id_list, process=str(process))
```
Comment (dagibbs22) [inline data_prep/mp_prep_other_inputs_one_off.py:843]: What is this line doing here? I see it was here previously-- just not sure what it's for.

File: removals/mp_gain_year_count_all_forest_types.py [hunk 8]
```diff
@@ -142,7 +142,7 @@ def mp_gain_year_count_all_forest_types(tile_id_list):
 
         # Creates gain year count tiles using only pixels that had neither loss nor gain pixels
         if cn.count == 96:
-            processes = 90   # 66 = 360 GB peak; 88 = 430 GB peak; 90 = 510 GB peak
+            processes = 50   # 66 = 360 GB peak; 88 = 430 GB peak; 90 = 510 GB peak
```
Comment (dagibbs22) [inline removals/mp_gain_year_count_all_forest_types.py:145]: Replace the comment with 50 = XXX GB peak.

File: removals/mp_gain_year_count_all_forest_types.py [hunk 9]
```diff
@@ -156,9 +156,9 @@ def mp_gain_year_count_all_forest_types(tile_id_list):
             pool.close()
             pool.join()
 
-        # Creates gain year count tiles using only pixels that had only gain
+        # Creates gain year count tiles using only pixels that had both loss and gain
         if cn.count == 96:
-            processes = 90   # 66 = 370 GB peak; 88 = 430 GB peak; 90 = 550 GB peak
+            processes = 50   # 66 = 370 GB peak; 88 = 430 GB peak; 90 = 550 GB peak
```
Comment (dagibbs22) [inline removals/mp_gain_year_count_all_forest_types.py:161]: Replace the comment with 50 = XXX GB peak.

File: removals/mp_gain_year_count_all_forest_types.py [hunk 10]
```diff
@@ -174,7 +174,7 @@ def mp_gain_year_count_all_forest_types(tile_id_list):
 
         # Combines the four above gain year count tiles for each Hansen tile into a single output tile
         if cn.count == 96:
-            processes = 84   # 28 processors = 220 GB peak; 62 = 470 GB peak; 78 = 600 GB peak; 80 = 620 GB peak; 84 = 630 GB peak
+            processes = 50   # 28 processors = 220 GB peak; 62 = 470 GB peak; 78 = 600 GB peak; 80 = 620 GB peak; 84 = 630 GB peak
```
Comment (dagibbs22) [inline removals/mp_gain_year_count_all_forest_types.py:177]: Replace the comment with 50 = XXX GB peak.

File: data_prep/mp_prep_other_inputs_one_off.py (hunk 7 above)
Comment (melrose-wri) [reply data_prep/mp_prep_other_inputs_one_off.py:843]: This is the main part of the script that is actually running the processing function we defined above and passing it the command line arguments/ constants, correct?

File: data_prep/mp_prep_other_inputs_annual.py [hunk 11]
```diff
@@ -5,8 +5,13 @@
 Hansenizing some removal factor standard deviation inputs, Hansenizing the European removal factors,
```
[the remark this answers is not in the document]
Comment (melrose-wri) [reply data_prep/mp_prep_other_inputs_annual.py]: Changes made in the following commit: b359551c7aa4405098e8f07ea7a582c6eddb963c

File: data_prep/mp_prep_other_inputs_annual.py (hunk 3 above)
Comment (melrose-wri) [reply data_prep/mp_prep_other_inputs_annual.py]: Changes made in the following commit: b359551c7aa4405098e8f07ea7a582c6eddb963c

File: data_prep/mp_prep_other_inputs_annual.py (hunk 4 above)
Comment (melrose-wri) [reply data_prep/mp_prep_other_inputs_annual.py]: Changes made in the following commit: b359551c7aa4405098e8f07ea7a582c6eddb963c

File: removals/mp_gain_year_count_all_forest_types.py (hunk 8 above)
Comment (melrose-wri) [reply removals/mp_gain_year_count_all_forest_types.py]: Changes made in the following commit: bcf63c1c802f8e9327579dfffe90b2654b52ff53

File: removals/mp_gain_year_count_all_forest_types.py (hunk 10 above)
Comment (melrose-wri) [reply removals/mp_gain_year_count_all_forest_types.py]: Changes made in the following commit: bcf63c1c802f8e9327579dfffe90b2654b52ff53

File: data_prep/mp_prep_other_inputs_one_off.py (hunk 7 above)
Comment (melrose-wri) [reply data_prep/mp_prep_other_inputs_one_off.py:843]: Changes made in the following commit: 064f1d8211bb345593bb2eca6e424b5b216a7475

File: data_prep/mp_prep_other_inputs_one_off.py (hunk 6 above)
Comment (melrose-wri) [reply data_prep/mp_prep_other_inputs_one_off.py]: Changes made in the following commit: 064f1d8211bb345593bb2eca6e424b5b216a7475

File: data_prep/mp_prep_other_inputs_one_off.py (hunk 5 above)
Comment (melrose-wri) [reply data_prep/mp_prep_other_inputs_one_off.py]: Changes made in the following commit: 064f1d8211bb345593bb2eca6e424b5b216a7475

File: data_prep/mp_prep_other_inputs_one_off.py [hunk 12]
```diff
@@ -137,6 +118,60 @@ def mp_prep_other_inputs(tile_id_list, process):
...
+            pool.join()
+
+        # Uploads output tiles to s3
+        if cn.NO_UPLOAD == False:
+            for i in range(0, len(output_dir_list)):
+                uu.upload_final_set(output_dir_list[i], output_pattern_list[i])
+
+    ##################################################################################################
+
+    if process == 'plantations' or process == 'all':
```
[the remark this answers is not in the document]
Comment (melrose-wri) [reply data_prep/mp_prep_other_inputs_one_off.py]: Changes made in the following commit: 064f1d8211bb345593bb2eca6e424b5b216a7475

**Review by dagibbs22 — APPROVED** [review]
Thanks for the changes. Looks good.

**Pull request merged**